### PR TITLE
Handle @typedef tag with missing type

### LINKED ISSUE #15852: Fix spelling for identifier in JSDoc
**TypeScript Version:**  nightly (2.4.0-dev.20170515)

**Code**

```ts
/**
 * @param abcdeg A parameter
 */
function f(abcdefg: number) {
    abcdeg;
}
```

**Expected behavior:**

Can fix spelling for either `abcdeg`.

**Actual behavior:**

Does not work in jsdoc.

### LINKED ISSUE #18637: Quick info on @typedef followed by declaration throws error
In the following example, try to get quick info on `A`.

```js
/**
 * @typedef A
 */
var x;
```

```
'quickinfo' request failed with error.
Error processing request. Cannot read property 'kind' of undefined
TypeError: Cannot read property 'kind' of undefined
    at getTypeFromTypeNode (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:25848:25)
    at getDeclaredTypeOfTypeAlias (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:23469:28)
    at Object.getDeclaredTypeOfSymbol (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:23592:24)
    at Object.getSymbolDisplayPartsDocumentationAndSymbolKind (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:69502:90)
    at Object.getQuickInfoAtPosition (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:76358:70)
    at IOSession.Session.getQuickInfoWorker (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:82196:62)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:81365:61)
    at C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:82741:88
    at IOSession.Session.executeWithRequestId (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:82732:28)
    at IOSession.Session.executeCommand (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:82741:33)
    at IOSession.Session.onMessage (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:82761:35)
    at Interface.<anonymous> (C:\Users\drosen\AppData\Roaming\npm\node_modules\typescript\lib\tsserver.js:83887:27)
    at emitOne (events.js:96:13)
    at Interface.emit (events.js:188:7)
    at Interface._onLine (readline.js:247:10)
    at Interface.<anonymous> (readline.js:384:12)
    at Array.forEach (native)
    at Interface._normalWrite (readline.js:383:11)
    at Socket.ondata (readline.js:107:10)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at Pipe.onread (net.js:551:20)
```

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #18637

I'll have to test tomorrow if starting to check jsdoc is hurting performance. But we'll need it to fix issue #15852.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -3507,6 +3507,10 @@
         "category": "Error",
         "code": 8020
     },
+    "JSDoc @typedef tag should either have a type annotation or be followed by @property or @member tags.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: `A JSDoc '@typedef' tag`...`by '@property' or '@member' tags.`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -22500,6 +22515,8 @@ namespace ts {
                 case SyntaxKind.ParenthesizedType:
                 case SyntaxKind.TypeOperator:
                     return checkSourceElement((<ParenthesizedTypeNode | TypeOperatorNode>node).type);
+                case SyntaxKind.JSDocTypedefTag:
+                    return checkJSDocTypedefTag(node as JSDocTypedefTag);
                 case SyntaxKind.JSDocComment:
                     return checkJSDocComment(node as JSDoc);
```
Comment (sandersn) [inline src/compiler/checker.ts]: This looks like it duplicates the new jsdoc-checking loop. Or perhaps this entry is now unneeded.

**Comment by sandersn** [comment]
To fix #15852, you'll also need binding, which means that you'll likely only be able to fix it for .js. Which is probably fine.

**Comment by andy-ms** [comment]
I ran performance tests and got little difference in compile times. (-0.16%, +0.68%, -0.99%, -1.01%, +0.97%, +1.60%)

**Comment by andy-ms** [comment]
@sandersn Good catch, there was redundant code checking jsdoc specifically for functions/methods.

**Pull request merged**